File: backend/modules/observability/application/convertor/trace/trajectory.go
// Copyright (c) 2025 coze-dev Authors
// SPDX-License-Identifier: Apache-2.0

package trace

import (
	dtotrajectory "github.com/coze-dev/coze-loop/backend/kitex_gen/coze/loop/trajectory"
	"github.com/coze-dev/coze-loop/backend/modules/observability/domain/trace/entity/loop_span"
)

func TrajectoriesDO2DTO(trajectories []*loop_span.Trajectory) []*dtotrajectory.Trajectory {
	if len(trajectories) == 0 {
		return nil
	}
	result := make([]*dtotrajectory.Trajectory, len(trajectories))
	for i, trajectory := range trajectories {
		result[i] = TrajectoryDO2DTO(trajectory)
	}
	return result
}

func TrajectoryDO2DTO(trajectory *loop_span.Trajectory) *dtotrajectory.Trajectory {
	if trajectory == nil {
		return nil
	}
	return &dtotrajectory.Trajectory{
		ID:         trajectory.ID,
		RootStep:   RootStepDO2DTO(trajectory.RootStep),
		AgentSteps: AgentStepsDO2DTO(trajectory.AgentSteps),
	}
}

func AgentStepsDO2DTO(steps []*loop_span.AgentStep) []*dtotrajectory.AgentStep {
	if len(steps) == 0 {
		return nil
	}
	result := make([]*dtotrajectory.AgentStep, len(steps))
	for i, step := range steps {
		result[i] = AgentStepDO2DTO(step)
	}
	return result
}

func AgentStepDO2DTO(step *loop_span.AgentStep) *dtotrajectory.AgentStep {
	if step == nil {
		return nil
	}
	return &dtotrajectory.AgentStep{
		ID:          step.ID,
		ParentID:    step.ParentID,
		Name:        step.Name,
		Input:       step.Input,
		Output:      step.Output,
		Steps:       StepsDO2DTO(step.Steps),
		Metadata:    step.Metadata,
		BasicInfo:   BasicInfoDO2DTO(step.BasicInfo),
		MetricsInfo: MetricsInfoDO2DTO(step.MetricsInfo),
	}
}

func StepsDO2DTO(steps []*loop_span.Step) []*dtotrajectory.Step {
	if len(steps) == 0 {
		return nil
	}
	result := make([]*dtotrajectory.Step, len(steps))
	for i, step := range steps {
		result[i] = StepDO2DTO(step)
	}
	return result
}

func StepDO2DTO(step *loop_span.Step) *dtotrajectory.Step {
	if step == nil {
		return nil
	}
	return &dtotrajectory.Step{
		ID:        step.ID,
		ParentID:  step.ParentID,
		Type:      step.Type,
		Name:      step.Name,
		Input:     step.Input,
		Output:    step.Output,
		ModelInfo: ModelInfoDO2DTO(step.ModelInfo),
		Metadata:  step.Metadata,
		BasicInfo: BasicInfoDO2DTO(step.BasicInfo),
	}
}

func ModelInfoDO2DTO(info *loop_span.ModelInfo) *dtotrajectory.ModelInfo {
	if info == nil {
		return nil
	}
	return &dtotrajectory.ModelInfo{
		InputTokens:               int64Ptr2int32Ptr(&info.InputTokens),
		OutputTokens:              int64Ptr2int32Ptr(&info.OutputTokens),
		LatencyFirstResp:          &info.LatencyFirstResp,
		ReasoningTokens:           int64Ptr2int32Ptr(&info.ReasoningTokens),
		InputReadCachedTokens:     int64Ptr2int32Ptr(&info.InputReadCachedTokens),
		InputCreationCachedTokens: int64Ptr2int32Ptr(&info.InputCreationCachedTokens),
	}
}

func int64Ptr2int32Ptr(src *int64) *int32 {
	if src == nil {
		return nil
	}
	result := int32(*src)
	return &result
}

func RootStepDO2DTO(step *loop_span.RootStep) *dtotrajectory.RootStep {
	if step == nil {
		return nil
	}
	return &dtotrajectory.RootStep{
		ID:          step.ID,
		Name:        step.Name,
		Input:       step.Input,
		Output:      step.Output,
		Metadata:    step.Metadata,
		BasicInfo:   BasicInfoDO2DTO(step.BasicInfo),
		MetricsInfo: MetricsInfoDO2DTO(step.MetricsInfo),
	}
}

func MetricsInfoDO2DTO(info *loop_span.MetricsInfo) *dtotrajectory.MetricsInfo {
	if info == nil {
		return nil
	}
	return &dtotrajectory.MetricsInfo{
		LlmDuration:        info.LlmDuration,
		ToolDuration:       info.ToolDuration,
		ToolErrors:         info.ToolErrors,
		ToolErrorRate:      info.ToolErrorRate,
		ModelErrors:        info.ModelErrors,
		ModelErrorRate:     info.ModelErrorRate,
		ToolStepProportion: info.ToolStepProportion,
		InputTokens:        info.InputTokens,
		OutputTokens:       info.OutputTokens,
	}
}

func BasicInfoDO2DTO(info *loop_span.BasicInfo) *dtotrajectory.BasicInfo {
	if info == nil {
		return nil
	}
	return &dtotrajectory.BasicInfo{
		StartedAt: &info.StartedAt,
		Duration:  &info.Duration,
		Error:     ErrorDO2DTO(info.Error),
	}
}

func ErrorDO2DTO(e *loop_span.Error) *dtotrajectory.Error {
	if e == nil {
		return nil
	}
	return &dtotrajectory.Error{
		Code: &e.Code,
		Msg:  &e.Msg,
	}
}

File: frontend/packages/loop-base/api-schema/src/api/idl/trajectory.ts
// Copyright (c) 2025 coze-dev Authors
// SPDX-License-Identifier: Apache-2.0
export interface Trajectory {
  /** trace_id */
  id?: string,
  /** 根节点，记录整个轨迹的信息 */
  root_step?: RootStep,
  /** agent step列表，记录轨迹中agent执行信息 */
  agent_steps?: AgentStep[],
}
export interface RootStep {
  /** 唯一ID，trace导入时取span_id */
  id?: string,
  /** name，trace导入时取span_name */
  name?: string,
  /** 输入 */
  input?: string,
  /** 输出 */
  output?: string,
  /**
   * 系统属性
   * 保留字段，可以承载业务自定义的属性
  */
  metadata?: {
    [key: string | number]: string
  },
  basic_info?: BasicInfo,
  metrics_info?: MetricsInfo,
}
export interface AgentStep {
  /**
   * 基础属性
   * 唯一ID，trace导入时取span_id
  */
  id?: string,
  /** 父ID， trace导入时取parent_span_id */
  parent_id?: string,
  /** name，trace导入时取span_name */
  name?: string,
  /** 输入 */
  input?: string,
  /** 输出 */
  output?: string,
  /** 子节点，agent执行内部经历了哪些步骤 */
  steps?: Step[],
  /**
   * 系统属性
   * 保留字段，可以承载业务自定义的属性
  */
  metadata?: {
    [key: string | number]: string
  },
  basic_info?: BasicInfo,
  metrics_info?: MetricsInfo,
}
export interface Step {
  /**
   * 基础属性
   * 唯一ID，trace导入时取span_id
  */
  id?: string,
  /** 父ID， trace导入时取parent_span_id */
  parent_id?: string,
  /** 类型 */
  type?: StepType,
  /** name，trace导入时取span_name */
  name?: string,
  /** 输入 */
  input?: string,
  /** 输出 */
  output?: string,
  /**
   * 各种类型补充信息
   * type=model时填充
  */
  model_info?: ModelInfo,
  /**
   * 系统属性
   * 保留字段，可以承载业务自定义的属性
  */
  metadata?: {
    [key: string | number]: string
  },
  basic_info?: BasicInfo,
}
export enum StepType {
  Agent = "agent",
  Model = "model",
  Tool = "tool",
}
export interface ModelInfo {
  input_tokens?: number,
  output_tokens?: number,
  /** 首包耗时，单位毫秒 */
  latency_first_resp?: string,
  reasoning_tokens?: number,
  input_read_cached_tokens?: number,
  input_creation_cached_tokens?: number,
}
export interface BasicInfo {
  /** 单位毫秒 */
  started_at?: string,
  /** 单位毫秒 */
  duration?: string,
  error?: Error,
}
export interface Error {
  code?: number,
  msg?: string,
}
export interface MetricsInfo {
  /** 单位毫秒 */
  llm_duration?: string,
  /** 单位毫秒 */
  tool_duration?: string,
  /** Tool错误分布，格式为：错误码-->list<ToolStepID> */
  tool_errors?: {
    [key: string | number]: string[]
  },
  /** Tool错误率 */
  tool_error_rate?: number,
  /** Model错误分布，格式为：错误码-->list<ModelStepID> */
  model_errors?: {
    [key: string | number]: string[]
  },
  /** Model错误率 */
  model_error_rate?: number,
  /** Tool Step占比(分母是总子Step) */
  tool_step_proportion?: number,
  /** 输入token数 */
  input_tokens?: number,
  /** 输出token数 */
  output_tokens?: number,
}